### PR TITLE
7.1 task added

### PR DESCRIPTION
**Pull request by p1q**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,52 +1,136 @@
...
+ * <p>Реалізувати свій ArrayList який імплементує інтерфейс List.</p>
  */
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size = 0;
+    private Object[] data;
+
+    public ArrayList(int initialCapacity) {
+        if (initialCapacity > 0) {
+            this.data = new Object[initialCapacity];
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
            data = new Object[initialCapacity];
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,52 +1,136 @@
...
+        } else {
+            throw new IllegalArgumentException("Wrong initial capacity: " + initialCapacity);
+        }
+    }
+
+    public ArrayList() {
+        this.data = new Object[DEFAULT_CAPACITY];
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
        data = new Object[DEFAULT_CAPACITY];
```
everywhere

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,52 +1,136 @@
...
+    }
+
+    private void ensureCapacity(int amount) {
+        if (data.length <= amount) {
+            data = Arrays.copyOf(data, data.length + amount + DEFAULT_CAPACITY);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: how many times we need to increase the size of the ArrayList?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,52 +1,136 @@
...
+        } else {
+            throw new IllegalArgumentException("Wrong initial capacity: " + initialCapacity);
+        }
+    }
+
+    public ArrayList() {
+        this.data = new Object[DEFAULT_CAPACITY];
+    }
+
+    private void ensureCapacity(int amount) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: do we need `int amount` here?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,52 +1,136 @@
...
+        throw new ArrayIndexOutOfBoundsException("The index is out of range: "
+                + index + ", the size is: " + this.size());
+    }
 
     @Override
     public void add(T value) {
-
+        ensureCapacity(size + 1);
+        size++;
+        data[size - 1] = value;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: size++
and then size - 1 ?
remove some of them

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,52 +1,136 @@
...
+            size++;
+            System.arraycopy(data, index, data, index + 1, size - index);
+            data[index] = value;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        ensureCapacity(size + 1);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: optional:
see how this method is implemented in the source code

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,52 +1,136 @@
...
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size) {
+            arrayIndexOutOfBoundsException(index);
+        }
+
+        T removedElement = (T) data[index];
+        if (index == size - 1) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: do we need this `if` here? maybe it is redundant?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 8]
```diff
@@ -1,52 +1,136 @@
...
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: sb?

File: src/main/java/core/basesyntax/ArrayList.java (hunk 3 above)
Comment (p1q) [reply src/main/java/core/basesyntax/ArrayList.java]: What do you mean "how many times"? One time

File: src/main/java/core/basesyntax/ArrayList.java (hunk 4 above)
Comment (p1q) [reply src/main/java/core/basesyntax/ArrayList.java]: I guess yes. How can we check whether we have needed capacity, if we don't know how much is needed capacity?

File: src/main/java/core/basesyntax/ArrayList.java (hunk 5 above)
Comment (p1q) [reply src/main/java/core/basesyntax/ArrayList.java]: But what's the problem?
size++; - we increase size
data[size - 1] = value; - and write the value to the end

File: src/main/java/core/basesyntax/ArrayList.java (hunk 6 above)
Comment (p1q) [reply src/main/java/core/basesyntax/ArrayList.java]: Ok, will check, thanks

File: src/main/java/core/basesyntax/ArrayList.java (hunk 3 above)
Comment (Sofasmile) [reply src/main/java/core/basesyntax/ArrayList.java]: в скільки разів потрібно збільшувати розмір масива?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 9]
```diff
@@ -1,52 +1,131 @@
...
+
+    private void ensureCapacity(int amount) {
+        if (data.length <= amount) {
+            data = Arrays.copyOf(data, data.length + amount + DEFAULT_CAPACITY);
+        }
+    }
+
+    private void arrayIndexOutOfBoundsException(int index) {
+        throw new ArrayIndexOutOfBoundsException("The index is out of range: "
+                + index + ", the size is: " + this.size());
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: remove this. everywhere

File: src/main/java/core/basesyntax/ArrayList.java (hunk 4 above)
Comment (Sofasmile) [reply src/main/java/core/basesyntax/ArrayList.java]: in 1.5

File: src/main/java/core/basesyntax/ArrayList.java [hunk 10]
```diff
@@ -1,52 +1,131 @@
...
+    }
 
     @Override
     public void add(T value) {
-
+        ensureCapacity(size + 1);
+        size++;
+        data[size - 1] = value;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
        data[size++] = value;
```and then remove         size++;

File: src/main/java/core/basesyntax/ArrayList.java [hunk 11]
```diff
@@ -1,52 +1,131 @@
...
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder listToString = new StringBuilder();
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: optional: we can use Arrays.toString()

File: src/main/java/core/basesyntax/ArrayList.java [hunk 12]
```diff
@@ -1,52 +1,125 @@
...
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+        if (index == size - 1) {
+            this.add(value);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
            add(value);
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 13]
```diff
@@ -1,52 +1,125 @@
...
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            if (size + 1 == data.length) {
+                data = Arrays.copyOf(data, data.length * 3 / 2);
+            }
+            this.add(list.get(i));
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
            add(list.get(i));
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 14]
```diff
@@ -1,52 +1,125 @@
...
+        return removedElement;
     }
 
     @Override
     public T remove(T t) {
+        if (t != null) {
+            for (int i = 0; i < size; i++) {
+                if (data[i].equals(t)) {
+                    T removedElement = (T) data[i];
+                    this.remove(i);
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
                    remove(i);
```

**Pull request closed without merging**